packages/runtime/container-runtime/src/dataStoreContexts.ts

```diff
@@ -88,15 +90,27 @@ import { FluidDataStoreContext, LocalFluidDataStoreContext } from "./dataStoreCo
         this._contexts.set(id, context);
 
         this.notBoundContexts.add(id);
-        this.prepDeferredContext(id);
+        this.ensureDeferred(id);
     }
 
     /**
-     * This returns a Deferred that will resolve when a context with the given id is bound,
-     * or added as remote from another client.
-     * @param id The id of the context to defer binding on
+     * Get a Promise that resolves to the requested context once it exists locally and is attached.
+     * e.g. If created locally, it must be bound, or if created remotely then it's fine as soon as it's sync'd in.
+     * Can also return undefined if the caller can't wait and the context isn't currently present and ready.
+     * @param id The id of the context to get
+     * @param wait If false, return undefined if the context isn't present and ready now. Otherwise return the Promise.
      */
-    public prepDeferredContext(id: string): Deferred<FluidDataStoreContext> {
+    public getBoundOrRemoted(id: string, wait: boolean): Promise<FluidDataStoreContext> | undefined {
```

Comment:
shouldn't this need to be async?

Reply:
It doesn't always return a promise (sometimes straight `undefined`). See other comment thread about the last commit's refactor here.

Reply:
i would put the undefined union into to promise, as it makes the api easier to use. It's not free, but it won't cause a perf issue here, it's only a problem if don't many many times on a critical path